lib/inputs/dynamodb-data.js

```diff
@@ -71,32 +70,58 @@ module.exports = class DynamodbData {
       return Promise.resolve([]);
     } else {
       const updates = uniqueKeys.map(k => [k, this.payloads[k], this.segments[k]]);
-      console.log(`Running ${uniqueKeys.length} DDB updates`);
-      const result = await dynamo.updateAll(updates);
-      const output = [{ dest: 'dynamodb', count: uniqueKeys.length }];
-
-      // send "new" segments onwards to kinesis
-      const formatted = result.success.map(r => this.format(r)).flat();
-      if (formatted.length) {
-        console.log(`Running ${formatted.length} kinesis puts`);
-        const count = await kinesis.put(formatted);
-        output.push({ count, dest: `kinesis:${kinesis.stream()}` });
+
+      logger.debug('Running DDB updates', { count: updates.length });
+      const [successCount, failureCount, loggedCount] = await this.updateAll(updates);
+      const output = [];
+      if (successCount) {
+        output.push({ dest: 'dynamodb', count: successCount });
+      }
+      if (loggedCount) {
+        output.push({ dest: 'kinesis', count: loggedCount });
       }
 
       // just throw an error if anything failed. successful records have already
       // been recorded in DDB, so retrying them won't cause duplicates.
-      if (result.failures.length > 0) {
-        const fails = result.failures.length;
-        const total = result.failures.length + result.success.length;
-        const msg = `DDB retrying for ${fails}/${total}`;
-        logger.warn(msg, { ddb: 'retrying', count: fails });
+      if (failureCount) {
+        const msg = `DDB retrying for ${failureCount}/${failureCount + successCount}`;
+        logger.warn(msg, { ddb: 'retrying', count: failureCount });
         throw new Error(msg);
       }
 
       return output;
     }
   }
 
+  // spin up N workers, and process the queue of updates
+  async updateAll(updates, concurrency = 25) {
+    let successCount = 0;
+    let failureCount = 0;
+    let loggedCount = 0;
+
+    const client = await dynamo.client();
+    const worker = async () => {
+      let args;
+      while ((args = updates.shift())) {
+        try {
+          const result = await dynamo.update.apply(this, args.concat([client]));
+          successCount++;
+          this.format(result).forEach(formatted => {
+            logger.info('impression', formatted);
```

Comment:
iirc, the dynamo batch inputs are idempotent (the batches are retried in entirety). Are these log statements also idempotent? Meaning that if the batch hits an error halfway through, it will be reprocessed and new log statements issued.

Reply:
I'm guessing you mean the _kinesis_ batch input?  Yes - those are idempotent, since we "lock" the segments we've already successfully processed in DDB.

And also yes - this log statement is idempotent as well.  If we succeeded in processing this key (`<listener_episode>.<digest>`) ... the segments will be locked on subsequent retries, and `this.format(result)` returns an empty array.